### PR TITLE
Fixes for transparency on nodata and ROI

### PR DESCRIPTION
**Pull request by aaime**

### CODE REVIEW

**Comment by dromagnoli** [comment]
Looks good to me.

File: modules/library/coverage/src/main/java/org/geotools/image/ImageWorker.java [hunk 1]
```diff
@@ -3962,11 +3963,19 @@ public ImageWorker affine(AffineTransform tx, Interpolation interpolation, doubl
...
                             setROI(newROI);
-                            paramBlk.add(warp).add(interpolation).add(bgValues).add(newROI);
+                            paramBlk.add(warp).add(interpolation).add(newROI);
                             if (oldNoData != null) {
                                 paramBlk.set(oldNoData, 4);
                             }
                         }
+                        
+                        // handle background values
+                        if(bgValues == null && sourceParamBlock.getNumParameters() > 2) {
```
Comment (dromagnoli) [inline modules/library/coverage/src/main/java/org/geotools/image/ImageWorker.java]: "if(condition)" should be "if (condition)" with a space between "if" and the "("
(Formatting rules)

File: modules/library/coverage/src/main/java/org/geotools/image/ImageWorker.java [hunk 2]
```diff
@@ -3962,11 +3963,19 @@ public ImageWorker affine(AffineTransform tx, Interpolation interpolation, doubl
...
+                            bgValues = (double[]) sourceParamBlock.getObjectParameter(2);
+                        }
+                        if(bgValues != null) {
+                            paramBlk.set(bgValues, 2);
```
Comment (dromagnoli) [inline modules/library/coverage/src/main/java/org/geotools/image/ImageWorker.java]: same as above

File: modules/library/coverage/src/main/java/org/geotools/image/ImageWorker.java [hunk 3]
```diff
@@ -4758,6 +4767,156 @@ public ImageWorker rangeLookup(Object rangeLookup) {
...
+    
+    /**
+     * Forces all NODATA pixels, as well as those outside of the ROI, to be transparent (expanding
+     * the color model as needed in order to make it so). In case the image has no ROI or no nodata,
+     * the method won't perform any change
+     */
+    public ImageWorker prepareForRendering() {
+        // anything to do?
+        ROI roi = getROI();
+        if(roi == null) {
```
Comment (dromagnoli) [inline modules/library/coverage/src/main/java/org/geotools/image/ImageWorker.java]: same here

File: modules/library/coverage/src/main/java/org/geotools/image/ImageWorker.java [hunk 4]
```diff
@@ -4758,6 +4767,156 @@ public ImageWorker rangeLookup(Object rangeLookup) {
...
+            Object roiCandidate = image.getProperty("roi");
+            if(roiCandidate instanceof ROI) {
```
Comment (dromagnoli) [inline modules/library/coverage/src/main/java/org/geotools/image/ImageWorker.java]: here

File: modules/library/coverage/src/main/java/org/geotools/image/ImageWorker.java [hunk 5]
```diff
@@ -4758,6 +4767,156 @@ public ImageWorker rangeLookup(Object rangeLookup) {
...
+                roi = (ROI) roiCandidate;
+            }
+        }
+        if(nodata == null && roi == null) {
```
Comment (dromagnoli) [inline modules/library/coverage/src/main/java/org/geotools/image/ImageWorker.java]: here

File: modules/library/coverage/src/main/java/org/geotools/image/ImageWorker.java [hunk 6]
```diff
@@ -4758,6 +4767,156 @@ public ImageWorker rangeLookup(Object rangeLookup) {
...
+                alphaChannels = new PlanarImage[] { PlanarImage.wrapRenderedImage(alpha) };
+            } 
+        } else if(cm instanceof ComponentColorModel) {
+            // convert to RGB if necessary
+            ComponentColorModel ccm = (ComponentColorModel) cm;
+            boolean hasAlpha = cm.hasAlpha();
+
+            // if we have a grayscale image see if we have to expand to RGB
+            if (ccm.getNumColorComponents() == 1) {
+                if((ccm.getTransferType() == DataBuffer.TYPE_DOUBLE ||
```
Comment (dromagnoli) [inline modules/library/coverage/src/main/java/org/geotools/image/ImageWorker.java]: here

File: modules/library/coverage/src/main/java/org/geotools/image/ImageWorker.java [hunk 7]
```diff
@@ -4758,6 +4767,156 @@ public ImageWorker rangeLookup(Object rangeLookup) {
...
+                    bgValues = new double[] { 0, 0, 0, 0 };
+                } 
+
+                // get back the ColorModel
+                cm = image.getColorModel();
+                ccm = (ComponentColorModel) cm;
+                hasAlpha = cm.hasAlpha();
+            }
+
+            if(bgValues == null) {
```
Comment (dromagnoli) [inline modules/library/coverage/src/main/java/org/geotools/image/ImageWorker.java]: here

File: modules/library/render/src/main/java/org/geotools/renderer/lite/gridcoverage2d/GridCoverageRenderer.java [hunk 8]
```diff
@@ -1204,14 +1200,22 @@ public void paint(final Graphics2D graphics, final GridCoverage2DReader gridCove
         }
     }
 
-    private void paintImage(final Graphics2D graphics, RenderedImage finalImage) {
+    private void paintImage(final Graphics2D graphics, RenderedImage inputImage) {
         final RenderingHints oldHints = graphics.getRenderingHints();
         graphics.setRenderingHints(this.hints);
+        
+        // nothing to do if the input image is null
+        if(inputImage == null) {
```
Comment (dromagnoli) [inline modules/library/render/src/main/java/org/geotools/renderer/lite/gridcoverage2d/GridCoverageRenderer.java]: same as above

**Comment by aaime** [comment]
Formatting fixed, merging

**Pull request merged**